lib/cache/collections.go

```diff
@@ -688,47 +706,58 @@ func (c *certAuthority) erase(ctx context.Context) error {
 	return nil
 }
 
-func (c *certAuthority) fetch(ctx context.Context) error {
-	if err := c.updateCertAuthorities(services.HostCA); err != nil {
-		return trace.Wrap(err)
+func (c *certAuthority) fetch(ctx context.Context) (apply func(ctx context.Context) error, err error) {
+	applyHostCAs, err := c.fetchCertAuthorities(services.HostCA)
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
-	if err := c.updateCertAuthorities(services.UserCA); err != nil {
-		return trace.Wrap(err)
+
+	applyUserCAs, err := c.fetchCertAuthorities(services.UserCA)
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
-	if err := c.updateCertAuthorities(services.JWTSigner); err != nil {
-		// DELETE IN: 5.1
-		//
-		// All clusters will support JWT signers in 5.1.
-		if strings.Contains(err.Error(), "authority type is not supported") {
-			return nil
-		}
-		return trace.Wrap(err)
+
+	applyJWTSigners, err := c.fetchCertAuthorities(services.JWTSigner)
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
-	return nil
+
+	return func(ctx context.Context) error {
+		if err := applyHostCAs(ctx); err != nil {
+			return trace.Wrap(err)
+		}
+		if err := applyUserCAs(ctx); err != nil {
+			return trace.Wrap(err)
+		}
+		return trace.Wrap(applyJWTSigners(ctx))
+	}, nil
 }
 
-func (c *certAuthority) updateCertAuthorities(caType services.CertAuthType) error {
+func (c *certAuthority) fetchCertAuthorities(caType services.CertAuthType) (apply func(ctx context.Context) error, err error) {
 	authorities, err := c.Trust.GetCertAuthorities(caType, c.watch.LoadSecrets, services.SkipValidation())
 	if err != nil {
 		// DELETE IN: 5.1
 		//
 		// All clusters will support JWT signers in 5.1.
-		if !strings.Contains(err.Error(), "authority type is not supported") {
-			return trace.Wrap(err)
+		if strings.Contains(err.Error(), "authority type is not supported") {
+			return func(ctx context.Context) error { return nil }, nil
 		}
+		return nil, trace.Wrap(err)
 	}
-	if err := c.trustCache.DeleteAllCertAuthorities(caType); err != nil {
-		if !trace.IsNotFound(err) {
-			return trace.Wrap(err)
+	return func(ctx context.Context) error {
+		if err := c.trustCache.DeleteAllCertAuthorities(caType); err != nil {
+			if !trace.IsNotFound(err) {
+				return trace.Wrap(err)
+			}
 		}
-	}
-	for _, resource := range authorities {
-		c.setTTL(resource)
-		if err := c.trustCache.UpsertCertAuthority(resource); err != nil {
-			return trace.Wrap(err)
+		for _, resource := range authorities {
+			c.setTTL(resource)
+			if err := c.trustCache.UpsertCertAuthority(resource); err != nil {
+				return trace.Wrap(err)
+			}
 		}
-	}
-	return nil
+		return nil
+	}, nil
```

Comment:
You didn't introduce this bug, but while testing this PR the below block is actually incorrect behavior, do you mind removing it?

```go
		// DELETE IN: 5.1.
		//
		// All clusters will support JWT signers in 5.1.
		if resource.GetType() == services.JWTSigner {
			return nil
		}
```